homeassistant/components/androidtv/media_player.py

```diff
@@ -168,180 +168,110 @@
 }
 
 
-def setup_androidtv(hass, config):
-    """Generate an ADB key (if needed) and load it."""
-    adbkey = config.get(CONF_ADBKEY, hass.config.path(STORAGE_DIR, "androidtv_adbkey"))
-    if CONF_ADB_SERVER_IP not in config:
-        # Use "adb_shell" (Python ADB implementation)
-        if not os.path.isfile(adbkey):
-            # Generate ADB key files
-            keygen(adbkey)
-
-        # Load the ADB key
-        signer = ADBPythonSync.load_adbkey(adbkey)
-        adb_log = f"using Python ADB implementation with adbkey='{adbkey}'"
-
-    else:
-        # Use "pure-python-adb" (communicate with ADB server)
-        signer = None
-        adb_log = f"using ADB server at {config[CONF_ADB_SERVER_IP]}:{config[CONF_ADB_SERVER_PORT]}"
-
-    return adbkey, signer, adb_log
-
-
-async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+async def async_setup_platform(
+    hass: HomeAssistant,
+    config: ConfigType,
+    async_add_entities: AddEntitiesCallback,
+    discovery_info=None,
+) -> None:
     """Set up the Android TV / Fire TV platform."""
-    hass.data.setdefault(ANDROIDTV_DOMAIN, {})
 
-    address = f"{config[CONF_HOST]}:{config[CONF_PORT]}"
+    host = config[CONF_HOST]
 
-    if address in hass.data[ANDROIDTV_DOMAIN]:
-        _LOGGER.warning("Platform already setup on %s, skipping", address)
-        return
-
-    adbkey, signer, adb_log = await hass.async_add_executor_job(
-        setup_androidtv, hass, config
-    )
+    # get main data
+    config_data = {
+        CONF_HOST: host,
+        CONF_DEVICE_CLASS: config.get(CONF_DEVICE_CLASS, DEFAULT_DEVICE_CLASS),
+        CONF_PORT: config.get(CONF_PORT, DEFAULT_PORT),
+    }
+    for key in (CONF_ADBKEY, CONF_ADB_SERVER_IP, CONF_ADB_SERVER_PORT, CONF_NAME):
+        if key in config:
+            config_data[key] = config[key]
+
+    # get options
+    config_options = {}
+    for key in (
+        CONF_APPS,
+        CONF_EXCLUDE_UNNAMED_APPS,
+        CONF_GET_SOURCES,
+        CONF_SCREENCAP,
+        CONF_STATE_DETECTION_RULES,
+        CONF_TURN_OFF_COMMAND,
+        CONF_TURN_ON_COMMAND,
+    ):
+        if key in config:
+            if key == CONF_STATE_DETECTION_RULES:
+                config_options[key] = json.dumps(config[key])
```

Comment:
There wasn't much feedback on this topic besides there isn't a need to json encode this since its going to be stored in json anyways so we end up with double encoded json in `.storage/core.config_entries`.

Reply:
You are suggesting to save this as JSON in `config_entry` and so convert the string field in json before save it in the `options`?

Reply:
I think you can drop the `json.dumps` and `json.loads` and store the data structure directly in the config entry.

Reply:
If we don't convert the JSON in string we will not be able to manage it using the options text box:
 
![image](https://user-images.githubusercontent.com/60491700/146660935-cd38e62f-d8ce-4504-a566-9c450cfc1353.png)

Reply:
I think I can improve a little bit this, creating a wizard similar to the one used for the apps list to manage rules for single apps. Please wait my next commit.

Reply:
I marked it as a draft.  When you are ready for another pass, please mark it ready for review

Reply:
I created the flow to configure the detection rules, looks more simple to manage in this way.
But I found a strange behavior that I don't think is related to this PR: when I call the method `self.async_create_entry` to save changed options, if I only change `apps` or `detection_rule` (that are managed with separated flow and saved as dict), not always new values are written in `.storage/core.config_entries`. It could be a bug or I'm missing something?

Reply:
Try making a `deepcopy` of the rules before modifying them. You might be updating the existing data structure which would prevent it from seeing it as new

Reply:
Of course, it was enough copy the 2 variable defined in init. Now it is ok!